mpi ghost communicator 3d

fankiat opened the pull request:
Fixes #4 

A few updates and notes to highlight:

- Deprecated the notion of `field_offset` and retain only `ghost_size`. We will set this as the max of ghost sizes for all kernels involved at the start of the simulation before initializing ghost communicator instance.
- I removed `n_values` from the parameters to be tested, since its more relevant when we are benchmarking later. Instead I introduced a `ghost_size` parameter for pytest, since that's what we need to test for ghost communication. The `n_values` are set directly in the test functions itself, and set accordingly to some values that works minimally to check code correctness. This is done in part also because the blocking exchange version of the code runs into some issues when array size is large (see next point for problem).
- Blocking communication falls into a deadlock-like situation when array size becomes too large (which honestly is not that large), similar to what is seen [here](https://stackoverflow.com/questions/21528275/mpi4py-hangs-when-trying-to-send-large-data). I tried the solutions discussed in the thread but to no avail. The most upvoted one suggested making the array into smaller chunks, which is almost a non-solution. I suspect it's an mpi4py library thing, and I am not too keen on investigating this in detail. In any case, we have a working version (see next point) and I am also somewhat inclined to deprecate the blocking version in the future, since it's expected slower anyway given its blocking nature for each send/recv instruction. I'm keeping the blocking version here for now, and we can deprecate it when we feel ready.
- Non-blocking communications does not have the problem above. Works fine for larger arrays and does not fail / deadlocks.

bhosale2 commented:
Some comments from my side:

>     * Deprecated the notion of `field_offset` and retain only `ghost_size`. We will set this as the max of ghost sizes for all kernels involved at the start of the simulation before initializing ghost communicator instance.

Sounds good!

>     * I removed `n_values` from the parameters to be tested, since its more relevant when we are benchmarking later. Instead I introduced a `ghost_size` parameter for pytest, since that's what we need to test for ghost communication. The `n_values` are set directly in the test functions itself, and set accordingly to some values that works minimally to check code correctness. This is done in part also because the blocking exchange version of the code runs into some issues when array size is large (see next point for problem).

 Let's reinstate the `n_values` and have 2 sets of them now say `[8, 16]`. The idea here is not to benchmark but test the correctness for varying grid sizes. If the blocking-type version causes errors for larger grid sizes, then that is a major problem in the library. To avoid this and ensure code correctness for different grid sizes, and given the early stage of development here, let's get rid of the blocking comms. now itself (no worries about breaking code), and no need to deprecate them later.

>     * Blocking communication falls into a deadlock-like situation when array size becomes too large (which honestly is not that large), similar to what is seen [here](https://stackoverflow.com/questions/21528275/mpi4py-hangs-when-trying-to-send-large-data). I tried the solutions discussed in the thread but to no avail. The most upvoted one suggested making the array into smaller chunks, which is almost a non-solution. I suspect it's an mpi4py library thing, and I am not too keen on investigating this in detail. In any case, we have a working version (see next point) and I am also somewhat inclined to deprecate the blocking version in the future, since it's expected slower anyway given its blocking nature for each send/recv instruction. I'm keeping the blocking version here for now, and we can deprecate it when we feel ready.

See my comment above, lets's remove it now and bring back the `n_values` range testing. Since repo is public, CI should be free and we don't need to be worried about the cost of testing `n_values=[8, 16]`, given the max time goes into installing the library anyways.

>     * Non-blocking communications does not have the problem above. Works fine for larger arrays and does not fail / deadlocks.

Nice, good investigation here, lets's just have this as the only working version then.

fankiat commented:
> @fankiat some comments from my side:
> 
> > ```
> > * Deprecated the notion of `field_offset` and retain only `ghost_size`. We will set this as the max of ghost sizes for all kernels involved at the start of the simulation before initializing ghost communicator instance.
> > ```
> 
> Sounds good!
> 
> > ```
> > * I removed `n_values` from the parameters to be tested, since its more relevant when we are benchmarking later. Instead I introduced a `ghost_size` parameter for pytest, since that's what we need to test for ghost communication. The `n_values` are set directly in the test functions itself, and set accordingly to some values that works minimally to check code correctness. This is done in part also because the blocking exchange version of the code runs into some issues when array size is large (see next point for problem).
> > ```
> 
> Let's reinstate the `n_values` and have 2 sets of them now say `[8, 16]`. The idea here is not to benchmark but test the correctness for varying grid sizes. If the blocking-type version causes errors for larger grid sizes, then that is a major problem in the library. To avoid this and ensure code correctness for different grid sizes, and given the early stage of development here, let's get rid of the blocking comms. now itself (no worries about breaking code), and no need to deprecate them later.
> 
> > ```
> > * Blocking communication falls into a deadlock-like situation when array size becomes too large (which honestly is not that large), similar to what is seen [here](https://stackoverflow.com/questions/21528275/mpi4py-hangs-when-trying-to-send-large-data). I tried the solutions discussed in the thread but to no avail. The most upvoted one suggested making the array into smaller chunks, which is almost a non-solution. I suspect it's an mpi4py library thing, and I am not too keen on investigating this in detail. In any case, we have a working version (see next point) and I am also somewhat inclined to deprecate the blocking version in the future, since it's expected slower anyway given its blocking nature for each send/recv instruction. I'm keeping the blocking version here for now, and we can deprecate it when we feel ready.
> > ```
> 
> See my comment above, lets's remove it now and bring back the `n_values` range testing. Since repo is public, CI should be free and we don't need to be worried about the cost of testing `n_values=[8, 16]`, given the max time goes into installing the library anyways.
> 
> > ```
> > * Non-blocking communications does not have the problem above. Works fine for larger arrays and does not fail / deadlocks.
> > ```
> 
> Nice, good investigation here, lets's just have this as the only working version then.

@bhosale2 sounds good. I have deprecated blocking exchange and now since we only have one method for exchanging ghost cells, I refactored the code to just call the `exchange()` function in the communicator.

For the `n_values` stuff, I agree with you mostly, but I took a slightly different approach, which tests for both large array sizes as well as problems of different `aspect_ratio`s. Please let me know what you think. I think this should be sufficient for testing size variation wise for robustness.

Thanks!

fankiat commented:
@bhosale2 found out I need to update the diffusion kernel code accordingly as well. So once all the CI stuff passes you can take a quick look and we can merge this.

bhosale2 commented on sopht_mpi/utils/mpi_utils_2d.py:
```diff
@@ -84,150 +84,55 @@ class MPIGhostCommunicator2D:
-    def __init__(self, ghost_size, field_offset, mpi_construct):
+    def __init__(self, ghost_size, mpi_construct):
         # extra width needed for kernel computation
+        assert ghost_size > 0, "ghost_size has to be > 0 for calling ghost comm."
```
same comment as before, lets raise ValueError

bhosale2 replied:
461c03a

bhosale2 commented on sopht_mpi/utils/mpi_utils_3d.py:
```diff
@@ -76,3 +76,188 @@ def __init__(
+
+    def __init__(self, ghost_size, mpi_construct):
+        # extra width needed for kernel computation
+        assert ghost_size > 0, "ghost_size has to be > 0 for calling ghost comm."
```
same comment as above.

bhosale2 replied:
461c03a

bhosale2 commented on tests/test_utils/test_mpi_utils_2d.py:
```diff
@@ -46,134 +46,53 @@ def test_mpi_field_io_gather_scatter(n_values, precision):
-@pytest.mark.parametrize("n_values", [16])
-def test_mpi_ghost_blocking_communication(n_values, precision):
+@pytest.mark.parametrize("rank_distribution", [(1, 0), (0, 1)])
+@pytest.mark.parametrize("aspect_ratio", [(1, 1), (1, 2), (2, 1), (1, 1)])
```
this is a good idea, maybe we should add this to the diffusion kernel as well?

fankiat replied:
Yeah that's a good idea. Implemented in [a0bce0a](https://github.com/fankiat/sopht-mpi/pull/5/commits/a0bce0a66151148448868c60bd9cc10fbf15d5b8)
Also good catch on the redundant aspect ratio. Thanks!

bhosale2 reviewed (changes requested):
Minor comments.

bhosale2 reviewed (approved):
LGTM. Will merge after CI passes, also the offset gathers and scatter functions need to be refactored since we don't have the concept of field offset anymore.
